firedrake/ensemble.py

```diff
@@ -0,0 +1,63 @@
+from pyop2.mpi import MPI
+
+__all__ = ("CommManager", )
+
+
+class CommManager(object):
```

Comment:
Let's bikeshed this name.  I picked `CommManager`, but I don't think it's good.  Do we want to call this `Ensemble`?  I think so.

Then, what attributes does it have?

In the normal way, it should have a `Ensemble.comm` attribute, but which communicator should this refer to.  In lunch discussion with @dham I think we concluded that it should be the "spatial communicator"?

Then, the global (input) communicator might be in the `gcomm` attribute.

Right now, we implement `allreduce`, we will also want `isend` and `irecv` probably (with the same semantics as the MPI equivalents).  These should take a source/target "ensemble rank", so this object should expose as `rank` the index of this member in the ensemble (that's `ecomm.rank` I think).

Did I miss things?

Reply:
It would be good to add isend and irecv.

Do we have to do them for this PR? I don't need them right now for my usecase, and don't feel too confident about writing them as I don't know the MPI lingo.